crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -346,6 +415,35 @@ impl<'w> EntityMut<'w> {
         self
     }
 
+    /// Adds a relation to an entity.
+    ///
+    /// This will overwrite previous values of the same relation to the same entity.
+    pub fn add_relation<R: Component>(&mut self, relation: R, target: Entity) -> &mut Self {
```

Comment:
Being able to query for relations is probably one of the next steps. Right now I'm of the opinion that something that should be forbidden initially is self referential relations ie. entities that have a relation with themselves as a target.

When making traversal APIs one of the things you want to support is getting parent components, self referential relations would easily cause UB. This is a problem I have in aery too and disallowing this kind of relation is how I fixed it.

The way Flecs deals with this is relations can be acyclic. Queries then check for cycles as they're traversing and panic if they're found. Self referential relations count as cycles. I'm not sure how long it would take to implement the same mechanism Flecs has but this kind of relation doesn't seem to allow for anything too useful and enabling them later shouldn't be too big a breaking change. Disallowing it initially however might make getting a query API off the ground easier.

Reply:
Good point- I'll add a check to panic if a self referential relationship is attempted.

Reply:
I think our next steps should probably be cleanup policies and dynamic components first, though. Without cleanup policies the ECS will trivially get into bad states constantly as entities despawn, and without dynamic components the entire API will have to be duplicate and separate from the builtin ones. I'm working on dynamic components now, and have decent progress. Will put up a PR if I feel happy with it.